src/Microsoft.Diagnostics.Monitoring.LogAnalytics/AuthenticatingHandler.cs

```diff
@@ -0,0 +1,72 @@
+﻿using Microsoft.Diagnostics.Tracing.Parsers.Clr;
+using Microsoft.Diagnostics.Tracing.Parsers.ClrPrivate;
+using Microsoft.Extensions.DependencyInjection;
+using System;
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Runtime.Serialization;
+using System.Security.Cryptography.X509Certificates;
+using System.Text;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Xml;
+
+namespace Microsoft.Diagnostics.Monitoring.LogAnalytics
+{
+    internal sealed class AuthenticationDelegatingHandler : DelegatingHandler
+    {
+        //TODO Storing a high value bearer token in plain text in memory
+        private AuthenticationHeaderValue _cachedBearerToken;
+        private readonly MetricsConfiguration _metricsConfiguration;
+
+        public AuthenticationDelegatingHandler(MetricsConfiguration configuration) : base(new HttpClientHandler())
+        {
+            _metricsConfiguration = configuration;
+        }
+
+        protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            if (request.Headers.Authorization == null)
+            {
+                if (_cachedBearerToken == null)
+                {
+                    _cachedBearerToken = await AcquireBearerToken(cancellationToken);
+                }
+                request.Headers.Authorization = _cachedBearerToken;
+            }
+
+            HttpResponseMessage response = await base.SendAsync(request, cancellationToken);
+            //Possible Token expired
+            if (response.StatusCode == System.Net.HttpStatusCode.Unauthorized)
```

Comment:
Should this be a loop?

Reply:
You mean like a retry handler? It's something we can add but i will do it separately.

---
In reply to: [387482772](https://github.com/wiktork/diagnostics/pull/1#discussion_r387482772) [](ancestors = 387482772)